src/compiler/factory/nodeFactory.ts

```diff
@@ -5275,7 +5275,7 @@ namespace ts {
             hasNoDefaultLib: boolean,
             libReferences: readonly FileReference[]
         ) {
-            const node = baseFactory.createBaseSourceFileNode(SyntaxKind.SourceFile) as Mutable<SourceFile>;
+            const node = (source.redirectInfo ? Object.create(source.redirectInfo.redirectTarget) : baseFactory.createBaseSourceFileNode(SyntaxKind.SourceFile)) as Mutable<SourceFile>;

```

Comment:
Is setting the prototype pretty typical in factory functions and whatnot @rbuckton?

Reply:
We generally don't, but apparently we use `Object.create` in program.ts for redirected source files.